port_forwarding.go

```diff
@@ -0,0 +1,386 @@
+)
+
+// sshPassword is a random password generated for the SSHD container.
+var sshPassword = uuid.NewString()
+
+// exposeHostPorts performs all the necessary steps to expose the host ports to the container, leveraging
+// the SSHD container to create the tunnel, and the container lifecycle hooks to manage the tunnel lifecycle.
+// At least one port must be provided to expose.
+// The steps are:
+// 1. Create a new SSHD container.
+// 2. Expose the host ports to the container after the container is ready.
+// 3. Close the SSH sessions before killing the container.
+func exposeHostPorts(ctx context.Context, req *ContainerRequest, p ...int) (ContainerLifecycleHooks, error) {
+	var sshdConnectHook ContainerLifecycleHooks
+
+	if len(p) == 0 {
+		return sshdConnectHook, fmt.Errorf("no ports to expose")
+	}
+
+	// Use the first network of the container to connect to the SSHD container.
+	var sshdFirstNetwork string
+	if len(req.Networks) > 0 {
+		sshdFirstNetwork = req.Networks[0]
+	} else if sshdFirstNetwork == "bridge" && len(req.Networks) > 1 {
+		sshdFirstNetwork = req.Networks[1]
+	}
+
+	opts := []ContainerCustomizer{}
+	if len(req.Networks) > 0 {
+		// get the first network of the container to connect the SSHD container to it.
+		nw, err := network.Get(ctx, sshdFirstNetwork)
+		if err != nil {
+			return sshdConnectHook, fmt.Errorf("failed to get the network: %w", err)
+		}
+
+		dockerNw := DockerNetwork{
+			ID:   nw.ID,
+			Name: nw.Name,
+		}
+
+		// WithNetwork reuses an already existing network, attaching the container to it.
+		// Finally it sets the network alias on that network to the given alias.
+		// TODO: Using an anonymous function to avoid cyclic dependencies with the network package.
+		withNetwork := func(aliases []string, nw *DockerNetwork) CustomizeRequestOption {
+			return func(req *GenericContainerRequest) {
+				networkName := nw.Name
+
+				// attaching to the network because it was created with success or it already existed.
+				req.Networks = append(req.Networks, networkName)
+
+				if req.NetworkAliases == nil {
+					req.NetworkAliases = make(map[string][]string)
+				}
+				req.NetworkAliases[networkName] = aliases
+			}
+		}
+
+		opts = append(opts, withNetwork([]string{hostInternal}, &dockerNw))
+	}
+
+	// start the SSHD container with the provided options
+	sshdContainer, err := newSshdContainer(ctx, opts...)
+	if err != nil {
+		return sshdConnectHook, fmt.Errorf("failed to create the SSH server: %w", err)
+	}
+
+	// IP in the first network of the container
+	sshdIP, err := sshdContainer.ContainerIP(context.Background())
+	if err != nil {
+		return sshdConnectHook, fmt.Errorf("failed to get IP for the SSHD container: %w", err)
+	}
+
+	if req.HostConfigModifier == nil {
+		req.HostConfigModifier = func(hostConfig *container.HostConfig) {}
+	}
+
+	// do not override the original HostConfigModifier
+	originalHCM := req.HostConfigModifier
+	req.HostConfigModifier = func(hostConfig *container.HostConfig) {
```

Comment:
nit: not sure about style in this project, but functions like this are easier to unit test if they are not anonymous.